src/main/scala/com/datasonnet/Mapper.scala

```diff
@@ -90,32 +93,43 @@ object Mapper {
     (name -> jsonnetLibrary)
   }
 
-  // TODO: will we require a header (or have a default header) on next release? I believe we should, as it is our last chance for a big required change, in which case
-  // the prepareForInput and prepareForOutput methods can be entirely removed!
-  // NOTE: this means there needs to be a match check between input and config, with JSON special cased as always allowed.
-  // this should be done by checking the plugin against the JSON plugin, as currently done.
-
-  def input(data: Document): Expr = {
-    // TODO once support for the header is integrated, if there is a header covering the input, call read.
-    // otherwise, call this method, which special cases the JSON plugin and treats every other identifier identically.
-    val json = DataFormatService.getInstance().prepareForInput(data)
-
-    Materializer.toExpr(json)
+  def input(name: String, data: Document, header: Header): Expr = {
+    val plugin = DataFormatService.getInstance().getPluginFor(data.mimeType)
+    if (plugin != null) {
+      val params = header.getInputParameters(name, data.mimeType)
+      checkParams(params, plugin.getReadParameters(), plugin.getPluginId)
+      val json = plugin.read(data.contents(), params.asInstanceOf[util.Map[String, AnyRef]])
+      Materializer.toExpr(json)
+    } else {
+      throw new IllegalArgumentException("The input mime type " + data.mimeType + " is not supported")
+    }
   }
 
-  def output(output: ujson.Value, mimeType: String): Document = {
-    // TODO once support for the header is integrated, if there is a header covering the output, call write.
-    // otherwise, call this method, which special cases the JSON plugin and treats every other identifier identically.
-    // NOTE: even if output is not specified, if a header is used, the default output provided should be JSON and any other must be provided explicitly.
-    val string = DataFormatService.getInstance().prepareForOutput(output, mimeType)
-    new StringDocument(string, mimeType)
+  def output(output: ujson.Value, mimeType: String, header: Header): Document = {
+    val plugin = DataFormatService.getInstance().getPluginFor(mimeType)
+    if (plugin != null) {
+      val params = header.getOutputParameters(mimeType)
+      checkParams(params, plugin.getWriteParameters(), plugin.getPluginId)
+      val str = plugin.write(output, params.asInstanceOf[util.Map[String, AnyRef]])
+      new StringDocument(str, mimeType)
+    } else {
+      throw new IllegalArgumentException("The output mime type " + mimeType + " is not supported")
+    }
   }
 
+  def checkParams(params: util.Map[_, _], supportedParams: util.Map[_, _], pluginId: String) = {
```

Comment:
This is all going to get interesting if ppl start offering competing plugins for the same mime types, with somewhat different parameters. I _think_ it's still a good thing: if they've got overlapping sets of parameters and you only use ones in the intersection, hopefully you don't care which is used? Anyways, that's a problem for future us if it ever comes up!